### PR TITLE
feat(router): added support for optional defend dispute api call and added evidence submission flow for checkout connector

### PR DESCRIPTION
**Pull request by sai-harsha-vardhan**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
1. added support for optional defend dispute api call in evidence submission flow if connector requires it.
2. added file upload and evidence submission flow for checkout connector.

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context
There is a use case where we have to do 2 api calls for evidence submission (for connectors like checkout, adyen)
1. Attach the evidence to dispute
2. Defend the dispute
Added support for optional second call to connector to defend the dispute

## How did you test it?
Tested Manually
image

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/api_models/src/enums.rs [hunk 1]
```diff
@@ -626,7 +626,7 @@ impl Connector {
         )
     }
     pub fn supports_file_storage_module(&self) -> bool {
-        matches!(self, Self::Stripe)
+        matches!(self, Self::Stripe | Self::Checkout)
```
Comment (Abhicodes-crypto) [inline crates/api_models/src/enums.rs:629]: Can this be populated from config ?

File: crates/router/src/connector/checkout.rs [hunk 2]
```diff
@@ -832,6 +828,322 @@ impl api::FileUpload for Checkout {
...
+        req: &types::SubmitEvidenceRouterData,
+        connectors: &settings::Connectors,
+    ) -> CustomResult<String, errors::ConnectorError> {
+        Ok(format!(
+            "{}{}{}{}",
+            self.base_url(connectors),
+            "disputes/",
+            req.request.connector_dispute_id,
+            "/evidence"
+        ))
```
Comment (Abhicodes-crypto) [inline crates/router/src/connector/checkout.rs:971]:
```suggestion
        Ok(format!(
            "{}disputes/{}/evidence"",
            self.base_url(connectors),
            req.request.connector_dispute_id,
        ))
```

File: crates/router/src/core/disputes.rs [hunk 3]
```diff
@@ -245,12 +245,52 @@ pub async fn submit_evidence(
...
+    )
+    .await
+    .change_context(errors::ApiErrorResponse::InternalServerError)
+    .attach_printable("Failed while calling defend dispute connector api")?;
+    let (dispute_status, connector_status) = match defend_response.response {
+        Ok(valid_defend_response) => Ok((
+            valid_defend_response.dispute_status,
+            valid_defend_response.connector_status,
+        )),
+        Err(err) => match err.message.as_ref() {
```
Comment (Abhicodes-crypto) [inline crates/router/src/core/disputes.rs:277]: Can `err.code` be used here ?

**Review by Abhicodes-crypto — CHANGES_REQUESTED** [review]
Minor changes requested , other than that LGTM

**Comment by Abhicodes-crypto** [comment]
Is there a limit on the file size that is uploaded to hyperswitch ?

**Review by Abhicodes-crypto — DISMISSED** [review]

File: crates/router/src/core/utils.rs [hunk 4]
```diff
@@ -400,3 +401,62 @@ pub async fn construct_upload_file_router_data<'a>(
...
+        payment_method_id: payment_attempt.payment_method_id.clone(),
+        address: PaymentAddress::default(),
+        auth_type: payment_attempt.authentication_type.unwrap_or_default(),
+        connector_meta_data: merchant_connector_account.get_metadata(),
+        amount_captured: payment_intent.amount_captured,
+        request: types::DefendDisputeRequestData {
+            dispute_id: dispute.dispute_id.clone(),
+            connector_dispute_id: dispute.connector_dispute_id.clone(),
+        },
+        response: Err(ErrorResponse::get_not_implemented()),
```
Comment (Narayanbhat166) [inline crates/router/src/core/utils.rs:455]: Is this error intentional? By default this should be `InternalServerError`

**Review by Narayanbhat166 — DISMISSED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Review by Abhicodes-crypto — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Pull request merged**